### ISSUE TITLE
Cannot use relative `require` inside of an FFI file

### ISSUE THREAD

**Issue by Pauan** [purescript/purescript#2343]
Here is an example which demonstrates the problem:

https://gist.github.com/Pauan/5c5cf6362c24b154b61910b8102a44f8

There are sometimes situations where it's useful to have an FFI file require another JS file. In the example, `Mutable.js` and `View.js` both require `Events.js`, in order to avoid code duplication. This is an internal implementation detail which is not exposed to PureScript.

However, this does not work, because the FFI files are copied into `output/${Module}/foreign.js`, and therefore the relative require will be relative to the `output/${Module}` directory, not the directory where the FFI file was originally.

So I propose that when `psc` is compiling an FFI file, it would **not** make a copy into `output/${Module}/foreign.js`, it would instead have `output/${Module}/index.js` point to the original file.

In other words, rather than `index.js` doing this:

```
var $foreign = require("./foreign");
```

It would instead do this:

```
var $foreign = require("../../path/to/file");
```

Is there any reason why `psc` makes a copy of the FFI file into the `output/${Module}` directory?

**Comment by paf31** [comment]
It is preferable, in my opinion, if the `output` directory is self-contained, except for globally-referenced modules pulled in by NPM.

If you need to share code between FFI modules, I'd recommend putting that shared code into its own (non-FFI) module and `require`ing it like a regular CommonJS module.

**Comment by Pauan** [comment]
I can understand the desire for cleanliness, however there are some downsides to that:

* Porting existing JS code to PureScript is harder.

* Having an application which is a mix of JS/PureScript code is *much* harder.

* It requires a separate CommonJS/Bower package even if it's just a simple file.

* It requires you to have an additional Rollup/Webpack build step, rather than relying entirely on `psc-bundle`.

Since the file already exists, I think it makes the most sense to just reference it, rather than making a copy, especially given the above benefits.

**Comment by Pauan** [comment]
@paf31 I thought of a compromise.

Rather than copying the FFI file into the `output` directory, instead create a symlink in the `output` directory that points to the FFI file.

* Tools which expect the FFI file to be in the `output` directory will continue to work.

* Tools which use `realpath` will detect that the FFI file and the symlink are the same, so they avoid duplicating the file, and they can resolve relative `require` correctly.

**Comment by paf31** [comment]
I don't really like the idea of relying on symlinks in foreign modules, but maybe this could be hidden behind a compiler flag.

But why not put the shared code in `node_modules` somewhere and require it using the module name?

**Comment by Pauan** [comment]
@paf31 Because they're small utility files that I didn't want to publish on npm.

JavaScript developers expect to be able to `require` JS files from inside JS files, this is a reasonable expectation.

Being able to `require` utility files is useful, and I think it will become even more useful with `rollup-plugin-purs`, which will automatically bundle them.

It's also useful for integrating PureScript into an already-existing JavaScript application (e.g. @wegry has a mix of JavaScript and PureScript code which uses Webpack and imports JS files from inside FFI files).

In the end I found a workaround, so it isn't super important to me to get this fixed, but I still think it would be a good idea.

**Comment by wegry** [comment]
@pauan Both of our workarounds (rollup plugin and module transpilation to disk) allow elision of the limitations and drawbacks of psc-bundle (which @paf31 has mentioned elsewhere he'd like to reduce the scope of).

Either or a combination of our approaches being adopted would remove this specific problem for users of the 'blessed' build stack. And from the look of it, at least from our conversations here, that seems likely.

**Comment by Pauan** [comment]
@wegry However, this problem is caused by `psc`, so it's unrelated to `psc-bundle`, therefore the same problem exists even with `rollup-plugin-purs`

**Comment by hdgarrood** [comment]
What about putting those JS files inside a directory which you then add to your NODE_PATH environment variable? Then you should be able to require them from FFI files.

**Comment by hdgarrood** [comment]
Alternatively I think using local paths in your package.json should work, and allows you to put the files in your node_modules directory without having to publish them to npm: https://docs.npmjs.com/files/package.json#local-paths

**Comment by wegry** [comment]
Ah, unless this has changed between 0.9 and 0.10, psc ignores CJS imports (requires).   All psc does with the FFI is parse, check exports, and force it to be ES5 CJS. I think the lack of resolution of imports is actually really handy, and it would be nice if it was a blessed convention post psc-bundle. 

You just import from where psc will put the generated CJS FFI module. require('b') needs to become require('b/foreign') as it currently stands. That is, if you're using rollup/webpack/systemjs.

Apologies for formatting, on my phone.

**Comment by Pauan** [comment]
@hdgarrood Using local packages is a decent workaround, though I really shouldn't have to workaround something which works correctly in pretty much every JS system *other* than PureScript.

Relative requires are an extremely common thing in JS-land, more so than npm requires, so JS developers expect it to work.

As far as I can see there isn't any benefit to copying the FFI files into the `output` directory.

If `psc` instead did *not* make copies of FFI files then everything would work correctly.

----

@wegry Of course there are workarounds, such as using `require("../Foo/foreign")`, but they are hacks which rely upon the implementation details for `psc`, and they shouldn't be necessary in the first place.

**Comment by hdgarrood** [comment]
I get that JS developers expect it to work, I just wanted to have it on the record that there are a decent selection of workarounds.

> If psc instead did not make copies of FFI files then everything would work correctly.

The only reason I'm reluctant about this idea is that it is not clear to me if/why this statement is true.

**Comment by Pauan** [comment]
@hdgarrood Fair enough.

Since tooling will already need to change for `psc-package`, perhaps now would be a good time to make this change as well.

In any case I agree that a proof of concept would be a good idea, but I'm not enthusiastic about creating a proof of concept if @paf31 is against this idea.

**Comment by paf31** [comment]
I'm not a fan of using relative paths to directories outside `output`.

**Comment by Pauan** [comment]
@paf31 Why aren't you a fan?

**Comment by paf31** [comment]
Because I don't want to couple the `src` and `output` directories.

**Comment by Pauan** [comment]
@paf31 I'm sorry, but I don't understand that at all. What "coupling" are you referring to, and what are the consequences of that coupling? Are there tools that break? Are there things which are more difficult? Are there assumptions which are violated?

**Comment by kRITZCREEK** [comment]
I'd also prefer we don't do this. The `output/` folder should contain everything one needs to use the compiled output imo. Having to chase more FFI files throughout the source tree is going to complicate any build tool that doesn't want to parse JS.

I think this should be solved by either moving the "required" code into node_modules/ or adding a build step in your own build chain that copies any `.js` files inside your source tree into the output directory/merging output and your other `.js` files into a `build/` directory.

**Comment by paf31** [comment]
> What "coupling" are you referring to

If I understand correctly, the build products under `output` would only make sense in the context of what's in `src`. I'd prefer the `output` directory be self-contained, or at least only rely on `node_modules`.

**Comment by LiamGoodacre** [comment]
Closing due to inactivity and that there have been many alternatives suggested in the previous comments.  Please reopen if there are further alternatives to be discussed.
